### PR TITLE
Add options to invocation context

### PR DESCRIPTION
**Pull request by ejizba**
Fixes https://github.com/Azure/azure-functions-nodejs-library/issues/52

The main goal is just to give people an easy way to inspect the configuration they used to register the function at the time of invocation.

### CODE REVIEW

File: types/InvocationContext.d.ts [hunk 1]
```diff
@@ -90,6 +90,12 @@ export declare class InvocationContext {
      * For example, this will be undefined for http and timer triggers because you can find that information on the request & timer object instead
      */
     triggerMetadata?: TriggerMetadata;
+
+    /**
+     * The options used when registering the function
+     * NOTE: This value may differ slightly from the original because it has been validated and defaults may have been explicitly added
+     */
+    options: EffectiveFunctionOptions;
```
Comment (hossam-nasr) [inline types/InvocationContext.d.ts:98]: `EffectiveFunctionOptions` has `extraInputs` and `extraOuputs` as optional, when in reality, they're always there, even if initialized to empty arrays (as I think they should). Can we have another type for this that expresses that?

File: src/InvocationModel.ts [hunk 2]
```diff
@@ -53,6 +54,7 @@ export class InvocationModel implements coreTypes.InvocationModel {
             retryContext: fromRpcRetryContext(req.retryContext),
             traceContext: fromRpcTraceContext(req.traceContext),
             triggerMetadata: fromRpcTriggerMetadata(req.triggerMetadata, this.#triggerType),
+            options: fromRpcBindings(this.#bindings),
```
Comment (hossam-nasr) [inline src/InvocationModel.ts:57]: My biggest issue with this implementation is that this now requires the `options` as input to the constuctor of `InvocationContext`, making the unit testing scenario harder. This code you have in the Wiki [here](https://github.com/Azure/azure-functions-nodejs-library/wiki/Azure-Functions-Node.js-Framework-v4) would no longer work:

```TS
const testInvocationContext = new InvocationContext({
  functionName: 'testFunctionName',
  invocationId: 'testInvocationId',
  logHandler: (_level, ...args) => console.log(...args)
});
```

Is there a way around this? Could we have the options object initialized inside the constructor, or set here after the `context = new InvocationContext({})` line?

File: types/InvocationContext.d.ts (hunk 1 above)
Comment (ejizba) [reply types/InvocationContext.d.ts:98]: Done

File: src/InvocationModel.ts (hunk 2 above)
Comment (ejizba) [reply src/InvocationModel.ts:57]: I assume this feedback applies to all properties on `InvocationContextInit`? I've updated so in testing scenarios you can just do `new InvocationContext()`

Comment (hossam-nasr) [reply src/InvocationModel.ts:57]: That's great! Yeah, I was gonna suggest that too but in a separate PR 😛

**Review by hossam-nasr — APPROVED** [review]
LGTM! Can we get a release of the library with these change so I can ingest it and make the necessary change on the SDK side?

**Pull request merged**